openhcl/page_pool_alloc/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+                for (i, state) in self.state.iter().enumerate() {
+                    resp.child(&i.to_string(), |req| match state {
+                        State::Free {
+                            base_pfn,
+                            pfn_bias,
+                            size_pages,
+                        } => {
+                            req.respond()
+                                .field("state", "free")
+                                .field("base_pfn", inspect::AsHex(base_pfn))
+                                .field("pfn_bias", inspect::AsHex(pfn_bias))
+                                .field("size_pages", inspect::AsHex(size_pages));
+                        }
+                        State::Allocated {
+                            base_pfn,
+                            pfn_bias,
+                            size_pages,
+                            device_id,
+                            tag,
+                        } => {
+                            req.respond()
+                                .field("state", "allocated")
+                                .field("base_pfn", inspect::AsHex(base_pfn))
+                                .field("pfn_bias", inspect::AsHex(pfn_bias))
+                                .field("size_pages", inspect::AsHex(size_pages))
+                                .field("device_id", self.device_ids[*device_id].clone())
+                                .field("tag", tag);
+                        }
+                    });
+                }
+            });
+    }
+}
+
+/// A handle for a page pool allocation. When dropped, the allocation is
+/// freed.
+#[derive(Debug)]
+pub struct PagePoolHandle {
+    inner: Arc<Mutex<PagePoolInner>>,
+    base_pfn: u64,
+    pfn_bias: u64,
+    size_pages: u64,
+}
+
+impl PagePoolHandle {
+    /// The base pfn (with bias) for this allocation.
+    pub fn base_pfn(&self) -> u64 {
+        self.base_pfn + self.pfn_bias
+    }
+
+    /// The base pfn without bias for this allocation.
+    pub fn base_pfn_without_bias(&self) -> u64 {
+        self.base_pfn
+    }
+
+    /// The number of 4K pages for this allocation.
+    pub fn size_pages(&self) -> u64 {
+        self.size_pages
+    }
+}
+
+impl Drop for PagePoolHandle {
+    fn drop(&mut self) {
+        let mut inner = self.inner.lock();
+
+        let index = inner
+            .state
+            .iter()
+            .position(|state| {
+                if let State::Allocated {
+                    base_pfn: base,
+                    pfn_bias: offset,
+                    size_pages: len,
+                    device_id: _,
+                    tag: _,
+                } = state
+                {
+                    *base == self.base_pfn && *offset == self.pfn_bias && *len == self.size_pages
+                } else {
+                    false
+                }
+            })
+            .expect("must find allocation");
+
+        inner.state[index] = State::Free {
```

Comment:
So we still don't do any merging on free.

I guess that's OK in practice since we just allocate during startup and don't really free.

Reply:
IIRC all the usage by devices expect contiguous pages, for queues and bounce buffers (unless I'm wrong and they're all just 4k each?). A lot of other callers just need a single 4K page, but I'd have to double check them all. 

And yeah, for free we don't _really_ expect any runtime allocations (the GET is the only one that might free but... we should probably fix that).

Reply:
NVMe queues are 4k each and this probably will never change. NVMe bounce buffers are bigger - 512k as of today, should be reduced once we investigate that.

> Do any callers rely on contiguous pages

yes, as discussed in NVMe save/restore, the assumption will be that pages are contiguous.